core/constraint-streams/src/test/java/ai/timefold/solver/constraint/streams/common/bi/AbstractBiConstraintStreamTest.java

```diff
@@ -3000,4 +3008,190 @@ public void joinerEqualsAndSameness() {
                 assertMatch(entity3, entity2));
     }
 
+    // ************************************************************************
+    // Node Sharing
+    // ************************************************************************
+
+    @Override
+    @TestTemplate
+    public void shareNodesWithSameInput() {
+        final List<String> failureList = new ArrayList<>();
+        buildScoreDirector(
+                TestdataSolution.buildSolutionDescriptor(),
+                factory -> {
+                    BiConstraintStream<TestdataEntity, TestdataEntity> base =
+                            factory.forEach(TestdataEntity.class).join(TestdataEntity.class);
+                    Function<TestdataEntity, TestdataEntity> function = a -> a;
+                    BiFunction<TestdataEntity, TestdataEntity, TestdataEntity> bifunction = (a, b) -> a;
+                    Function<TestdataEntity, Iterable<String>> iterableFunction = a -> Collections.emptyList();
+                    BiPredicate<TestdataEntity, TestdataEntity> bipredicate = (a, b) -> true;
+                    TriPredicate<TestdataEntity, TestdataEntity, TestdataEntity> tripredicate = (a, b, c) -> true;
+
+                    if (base.join(TestdataEntity.class) != base.join(TestdataEntity.class)) {
+                        failureList.add("join(no-joiners)");
+                    }
```

Comment:
This is needlessly complicated. Since you never need to call those constraint streams, you don't need `buildScoreDirector` or `failureList`.

Let's change the test to look like this:

    Assertions.assertThat(base.join(...))
        .as(...)
        .isSameAs(base.join(...)

Maybe you'll need to figure out how to get a ConstraintFactory instance, but it will be better than using this `failureList` hack. It is also part of my reasoning for having this in a new test class, because this will require new approaches and generally be unlike the other tests in the current class.

Reply:
In fact if each such assertions becomes its own well named test method, even better, because we will be able to introduce an interface for all 4 cardinalities, enforcing the kinds of things we want to test in each.